File: i2pchat/storage/sealed_json.py
"""At-rest JSON blobs wrapped in NaCl SecretBox (identity-keyed).

On-disk format:
  4 bytes  — magic
  2 bytes  — wrap version (big-endian uint16, currently 1)
  32 bytes — salt
  rest     — SecretBox(JSON)

Legacy plaintext UTF-8 JSON is still read and re-encrypted on the next save
when an identity key is available.
"""

from __future__ import annotations

import json
import logging
import secrets
import struct
from typing import Any

from i2pchat import crypto
from i2pchat.storage.blindbox_state import atomic_write_bytes, atomic_write_json

logger = logging.getLogger("i2pchat.storage.sealed_json")

SEALED_JSON_ENC_VERSION = 1
SEALED_JSON_SALT_SIZE = 32
SEALED_JSON_HEADER_SIZE = 4 + 2 + SEALED_JSON_SALT_SIZE


def _derive_file_key(identity_key: bytes, salt: bytes, domain: bytes) -> bytes:
    prk = crypto.hkdf_extract(domain, identity_key)
    profile_key = crypto.hkdf_expand(prk, domain + b"|profile-key", 32)
    prk2 = crypto.hkdf_extract(salt, profile_key)
    return crypto.hkdf_expand(prk2, domain + b"|file-key", 32)


def _read_existing_salt(path: str, magic: bytes) -> bytes | None:
    try:
        with open(path, "rb") as handle:
            header = handle.read(SEALED_JSON_HEADER_SIZE)
    except OSError:
        return None
    if len(header) < SEALED_JSON_HEADER_SIZE or header[:4] != magic:
        return None
    return header[6:SEALED_JSON_HEADER_SIZE]


def is_sealed_json_file(path: str, magic: bytes) -> bool:
    return _read_existing_salt(path, magic) is not None


def write_sealed_json(
    path: str,
    payload: dict[str, Any],
    *,
    identity_key: bytes | None,
    magic: bytes,
    domain: bytes,
) -> None:
    existing_encrypted = is_sealed_json_file(path, magic)
    if not identity_key or not crypto.NACL_AVAILABLE:
        if existing_encrypted:
            logger.warning(
                "refusing to overwrite encrypted %s without an identity key",
                path,
            )
            return
        if identity_key and not crypto.NACL_AVAILABLE:
            logger.warning(
                "PyNaCl unavailable — writing %s without at-rest encryption",
                path,
            )
        atomic_write_json(path, payload)
        return
    salt = _read_existing_salt(path, magic) or secrets.token_bytes(SEALED_JSON_SALT_SIZE)
    file_key = _derive_file_key(identity_key, salt, domain)
    plaintext = json.dumps(payload, ensure_ascii=True, separators=(",", ":")).encode(
        "utf-8"
    )
    ciphertext = crypto.encrypt_message(file_key, plaintext)
    header = magic + struct.pack(">H", SEALED_JSON_ENC_VERSION) + salt
    atomic_write_bytes(path, header + ciphertext)


def read_sealed_json(
    path: str,
    *,
    identity_key: bytes | None,
    magic: bytes,
    domain: bytes,
) -> dict[str, Any]:
    with open(path, "rb") as handle:
        raw = handle.read()
    if raw[:4] != magic:
        return json.loads(raw.decode("utf-8"))
    if len(raw) < SEALED_JSON_HEADER_SIZE:
        raise ValueError("Sealed JSON record truncated")
    if not identity_key or not crypto.NACL_AVAILABLE:
        raise ValueError("Record is encrypted but no identity key is available")
    version = struct.unpack(">H", raw[4:6])[0]
    if version != SEALED_JSON_ENC_VERSION:
        raise ValueError(f"Unsupported sealed JSON version {version}")
    salt = raw[6:SEALED_JSON_HEADER_SIZE]
    ciphertext = raw[SEALED_JSON_HEADER_SIZE:]
    file_key = _derive_file_key(identity_key, salt, domain)
    plaintext = crypto.decrypt_message(file_key, ciphertext)
    if plaintext is None:
        raise ValueError("Sealed JSON decryption failed (wrong key or tampered)")
    payload = json.loads(plaintext.decode("utf-8"))
    if not isinstance(payload, dict):
        raise ValueError("Sealed JSON payload must be an object")
    return payload

File: cpp/core/src/storage/sealed_json.cpp
#include "i2pchat/storage/sealed_json.hpp"

#include <algorithm>
#include <fstream>

#include "i2pchat/crypto.hpp"
#include "i2pchat/storage/atomic_write.hpp"

namespace i2pchat::storage {
namespace {

/// Read just the header, without pulling a potentially large file into memory.
std::optional<Bytes> read_header(const std::filesystem::path& path,
                                 std::string_view magic) {
    std::ifstream stream(path, std::ios::binary);
    if (!stream) {
        return std::nullopt;
    }
    Bytes header(kSealedJsonHeaderSize);
    stream.read(reinterpret_cast<char*>(header.data()),
                static_cast<std::streamsize>(header.size()));
    if (stream.gcount() != static_cast<std::streamsize>(kSealedJsonHeaderSize)) {
        return std::nullopt;
    }
    if (!std::equal(magic.begin(), magic.end(), header.begin())) {
        return std::nullopt;
    }
    return header;
}

}  // namespace

Bytes derive_sealed_profile_key(ByteView identity_key, std::string_view domain) {
    const Bytes prk = crypto::hkdf_extract(as_bytes(domain), identity_key);
    const std::string info = std::string(domain) + "|profile-key";
    return crypto::hkdf_expand(ByteView(prk), as_bytes(info), 32);
}

Bytes derive_sealed_file_key(ByteView identity_key, ByteView salt,
                             std::string_view domain, std::string_view scope) {
    const Bytes profile_key = derive_sealed_profile_key(identity_key, domain);
    const Bytes prk = crypto::hkdf_extract(salt, ByteView(profile_key));
    std::string info = std::string(domain) + "|file-key";
    if (!scope.empty()) {
        info += "|";
        info += scope;
    }
    return crypto::hkdf_expand(ByteView(prk), as_bytes(info), 32);
}

bool is_sealed_json_file(const std::filesystem::path& path, std::string_view magic) {
    return read_header(path, magic).has_value();
}

std::string serialize_sealed_payload(const nlohmann::json& payload) {
    return payload.dump(-1, ' ', /*ensure_ascii=*/true);
}

nlohmann::json read_sealed_json(const std::filesystem::path& path,
                                std::optional<ByteView> identity_key,
                                const SealedJsonFormat& format) {
    const Bytes raw = read_file(path);
    const std::string_view magic = format.magic;

    const bool sealed = raw.size() >= magic.size() &&
                        std::equal(magic.begin(), magic.end(), raw.begin());
    if (!sealed) {
        // Legacy plaintext, written before at-rest encryption existed.
        return nlohmann::json::parse(to_string(ByteView(raw)));
    }
    if (raw.size() < kSealedJsonHeaderSize) {
        throw SealedJsonError("Sealed JSON record truncated");
    }
    if (!identity_key.has_value()) {
        throw SealedJsonError("Record is encrypted but no identity key is available");
    }

    const std::uint16_t version = read_u16_be(ByteView(raw).subspan(4, 2));
    if (version != format.version) {
        throw SealedJsonError("Unsupported sealed JSON version " +
                              std::to_string(version));
    }

    const ByteView salt = ByteView(raw).subspan(6, kSealedJsonSaltSize);
    const ByteView ciphertext = ByteView(raw).subspan(kSealedJsonHeaderSize);
    const Bytes file_key =
        derive_sealed_file_key(*identity_key, salt, format.domain, format.scope);

    const std::optional<Bytes> plaintext =
        crypto::decrypt_message(ByteView(file_key), ciphertext);
    if (!plaintext.has_value()) {
        throw SealedJsonError("Sealed JSON decryption failed (wrong key or tampered)");
    }

    nlohmann::json payload = nlohmann::json::parse(to_string(ByteView(*plaintext)));
    if (!payload.is_object()) {
        throw SealedJsonError("Sealed JSON payload must be an object");
    }
    return payload;
}

void write_sealed_json(const std::filesystem::path& path, const nlohmann::json& payload,
                       std::optional<ByteView> identity_key,
                       const SealedJsonFormat& format) {
    const std::optional<Bytes> existing_header = read_header(path, format.magic);

    if (!identity_key.has_value()) {
        if (existing_header.has_value()) {
            // Overwriting sealed data with plaintext would quietly strip a
            // user's at-rest encryption, so refuse instead.
            throw SealedJsonError(
                "Refusing to overwrite an encrypted record without an identity key");
        }
        atomic_write_json(path, payload);
        return;
    }

    // Reuse the file's salt when it already has one: the reference
    // implementation keeps it stable for the life of the file.
    Bytes salt;
    if (existing_header.has_value()) {
        salt.assign(existing_header->begin() + 6, existing_header->end());
    } else {
        salt = crypto::random_bytes(kSealedJsonSaltSize);
    }

    const Bytes file_key = derive_sealed_file_key(*identity_key, ByteView(salt),
                                                  format.domain, format.scope);
    const std::string plaintext = serialize_sealed_payload(payload);
    const Bytes ciphertext =
        crypto::encrypt_message(ByteView(file_key), as_bytes(plaintext));

    Bytes blob;
    blob.reserve(kSealedJsonHeaderSize + ciphertext.size());
    append(blob, format.magic);
    append_u16_be(blob, format.version);
    append(blob, ByteView(salt));
    append(blob, ByteView(ciphertext));
    atomic_write_bytes(path, ByteView(blob));
}

}  // namespace i2pchat::storage
